src/sync_user.hpp

```diff
@@ -0,0 +1,118 @@
+	// The session will be returned synchronously. However, the 'bind session callback' might
+	// also be invoked, and the session might not actually be 'active' until the network response
+	// triggered by that callback is received and the appropriate `bind()` API called.
+	//
+	// This can be used in one of two ways: either automatically by `RealmCoordinator` in the
+	// process of opening a Realm configured for sync, or directly by the binding in order to
+	// get a 'stand-alone' session for whatever reason. 
+	std::shared_ptr<SyncSession> get_session(const std::string& path, const SyncConfig& config);
+
+	// Get a session, but only if it already exists and is active.
+    std::shared_ptr<SyncSession> get_existing_active_session(const std::string& path) const;
+
+	// Log the user out and mark it as invalid. This will also close its associated Sessions.
+	void log_out();
+
+	// Whether the user was configured as an 'admin user' (directly uses its user token
+	// to open Realms).
+	bool is_admin() const
+	{
+		return m_is_admin;
+	}
+
+	std::string identity() const
+	{
+		return m_identity;
+	}
+
+	State state() const
+	{
+		return m_state;
+	}
+
+	// The auth server URL. Bindings should set this appropriately when they retrieve
+	// instances of `SyncUser`s.
+	// FIXME: once the new token system is implemented, this can be removed completely.
+	std::string server_url;
+
+private:
+	State m_state;
+	
+	// Whether the user is an 'admin' user. Admin users use the admin tokens they were
+	// configured with to directly open sessions, and do not make network requests.
+	bool m_is_admin;
+	// The token for the user. Managed by the SyncManager APIs that vend out users.
+	std::string m_user_token;
+	// Set by the server. The unique ID of the user account on the Realm Object Server.
+	std::string m_identity;
+
+	// Protects m_active_sessions and m_inactive_sessions
+    mutable std::mutex m_session_mutex;
+
+    std::shared_ptr<SyncSession> get_existing_active_session_locked(const std::string& path) const;
+    std::unique_ptr<SyncSession> get_existing_inactive_session_locked(const std::string& path);
+
+    void dropped_last_reference_to_session(SyncSession*);
+
+    // Stop tracking the session for the given path if it is inactive.
+    // No-op if the session is either still active or in the active sessions list
+    // due to someone holding a strong reference to it.
+    void unregister_session(const std::string& path);
+
+    // Active sessions are sessions which the client code holds a strong
+    // reference to. When the last strong reference is released, the session is
+    // moved to inactive sessions. Inactive sessions are promoted back to active
+    // sessions until the session itself calls unregister_session to remove
+    // itself from inactive sessions once it's done with whatever async cleanup
+    // it needs to do.
+    std::unordered_map<std::string, std::weak_ptr<SyncSession>> m_active_sessions;
```

Comment:
> Why should two users not be able to open the same Realm? That seems like a pretty common use case for shared data.

There's no reason for two users to open the same Realm on disk. They can certainly open the same object server URL though.

> A given Realm is tied to a single user (and its path reflects this fact), so it should by definition be impossible for two users to open the same Realm.

What makes this impossible? AFAICT there's nothing stopping me from creating two separate `SyncUser`s and asking each to give me a session for the same path on disk.